src/cdomains/lval.ml

```diff
@@ -256,6 +256,18 @@ struct
   let to_string = function
     | StrPtr (Some x) -> Some x
     | _        -> None
+  let to_n_string n = function
+    | StrPtr (Some x) -> 
+      if n > String.length x then
+        Some x
+      else if n < 0 then
+        None
+      else
+        Some (String.sub x 0 n)
+    | _ -> None
+  let to_string_length = function
+    | StrPtr (Some x) -> Some (String.length x)
+    | _ -> None
```

Comment:
I suspect this might not be so easy because C strings and OCaml strings are different. In particular a string literal `"hello\0world"` would have `strlen` 5, but the OCaml string has length 11.
The same probably applies to other length-related functionality as well.

Maybe CIL has some helper functions for manipulating strings with C semantics?

Reply:
That's absolutely right, again thank you for thinking about this case! I couldn't find a CIL function that would help me in this case in a quick search, but I'll look at it again. Otherwise, `String.split_on_char '\x00' x` and taking the first element of the resulting list would do the trick.

Reply:
It is now fixed for `strlen`, but I think it still needs to be fixed for some of the other functions. I think it matters for `strstr` and `strcmp` as well. Right now they're just performing the operations on OCaml strings, but all of those operations need to be performed with C semantics, so only up to the NUL byte considered when checking for inclusion etc.

I'm wondering if a better solution is to maybe just do the stripping initially, so it doesn't have to be done for every operation individually. Or maybe CIL is already doing that as well?
I haven't checked, but it would be good to have tests for these things. If sv-benchmarks doesn't contain such cases, it would be nice to also contribute those because this detail is easy to get wrong and become unsound.

Reply:
My first assumption was that CIL would cut the string at the first NULL byte, that's why I didn't do the stripping. After your comment, I tested this and found that's not the case.

As you're right that this also concerns `strstr` and `str(n)cmp`, I introduced a new function `to_c_string` performing the stripping which I now use accross all my other functions. I didn't change `to_string` as some parts of the base analysis not related to my PR rely on it.

I added a new *02-string_literals_with_null.c* test to verify this edge case is now correctly covered.